fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/postgresql/PostgreSqlParameterNamesDAO.java

```diff
@@ -6,75 +6,101 @@
-        Integer result = getParameterId(parameterName);
+        final String METHODNAME = "readOrAddParameterNameId";
+        log.entering(CLASSNAME, METHODNAME);
 
-        // Create the resource if we don't have it already (set by the continue handler)
-        if (result == null) {
-            try {
-                result = fhirRefSequenceDAO.nextValue();
+        int parameterNameId;
+        String currentSchema;
+        String stmtString;
+        String errMsg = "Failure storing search parameter name id: name=" + parameterName;
+        long dbCallStartTime;
+        double dbCallDuration;
 
-                final String INS = "INSERT INTO parameter_names (parameter_name_id, parameter_name) VALUES (?, ?)";
-                try (PreparedStatement stmt = getConnection().prepareStatement(INS)) {
-                    // bind parameters
-                    stmt.setInt(1, result);
-                    stmt.setString(2, parameterName);
-                    stmt.executeUpdate();
+        try {
+            // TODO: schema should be known by application. Fix to avoid an extra round-trip.
+            currentSchema = getConnection().getSchema().trim();
+            stmtString = String.format(SQL_CALL_ADD_PARAMETER_NAME, currentSchema);
+            try (CallableStatement stmt = getConnection().prepareCall(stmtString)) {
+                stmt.setString(1, parameterName);
+                stmt.registerOutParameter(2, Types.INTEGER);
+                dbCallStartTime = System.nanoTime();
+                stmt.execute();
+                dbCallDuration = (System.nanoTime()-dbCallStartTime)/1e6;
+                if (log.isLoggable(Level.FINE)) {
+                        log.fine("DB read/store parameter name id complete. executionTime=" + dbCallDuration + "ms");
                 }
-            } catch (SQLException e) {
-                throw new FHIRPersistenceDataAccessException("Error while getting or inserting parameterName '" + parameterName + "'", e);
+                parameterNameId = stmt.getInt(2);
             }
         }
-
-        // cannot be null, so safe to return as an int
-        return result;
+        catch (Throwable e) {
+            throw new FHIRPersistenceDataAccessException(errMsg,e);
+        }
+        finally {
+            this.cleanup(null, getConnection());
+            log.exiting(CLASSNAME, METHODNAME);
+        }
+        return parameterNameId;
     }
 
     /**
-     * Read the id for the named type
-     * @param parameterName
-     * @return the database id, or null if the named record is not found
-     * @throws FHIRPersistenceDataAccessException
+     * Closes the passed PreparedStatement and Connection objects.
+     * @param stmt
+     * @param connection
      */
-    protected Integer getParameterId(String parameterName) throws FHIRPersistenceDataAccessException {
-        Integer result;
+    protected void cleanup(PreparedStatement stmt, Connection connection)  {
```

Comment:
once again, it's a pool, this is a confusing method, when is it called, why is it called, why is it used. 
The preparedstatement should be autocloseable. 

In fact in this case, I think it should be removed all together.

Reply:
this doesn't really close the connection, just return the connection to the pool. it's the same in db2 codes. but seems no harm to just drop it here.